File: 프로그래머스/lv2/이재표/교점에 별 만들기.java
import java.util.*;

class Solution {
    public String[] solution(int[][] line) {
        Set<int[]> set = new HashSet<>();

        for (int i = 0; i < line.length; i++) {
            for (int j = i + 1; j < line.length; j++) {
                int A = line[i][0];
                int B = line[i][1];
                int E = line[i][2];

                int C = line[j][0];
                int D = line[j][1];
                int F = line[j][2];

                long under =(long) A * D - (long)B * C;
                if (under == 0) continue; 

                long xi = (long)B * F - (long)E * D;
                long yi =(long) E * C - (long)A * F;

                if (xi % under != 0 || yi % under != 0) continue; 

                int x =(int)( xi / under);
                int y =(int)( yi / under);

                set.add(new int[]{x, y});
            }
        }

        int minX = Integer.MAX_VALUE;
        int maxX = Integer.MIN_VALUE;
        int minY = Integer.MAX_VALUE;
        int maxY = Integer.MIN_VALUE;

        for (int[] p : set) {
            minX = Math.min(minX, p[0]);
            maxX = Math.max(maxX, p[0]);
            minY = Math.min(minY, p[1]);
            maxY = Math.max(maxY, p[1]);
        }

        int width = maxX - minX + 1;
        int height = maxY - minY + 1;

        char[][] grid = new char[height][width];
        for (char[] row : grid) {
            Arrays.fill(row, '.');
        }

        for (int[] p : set) {
            grid[maxY - p[1]][p[0] - minX] = '*';
        }

        String[] answer = new String[height];
        for (int i = 0; i < height; i++) {
            answer[i] = new String(grid[i]);
        }

        return answer;
    }
}

File: 프로그래머스/lv2/김유현/교점에 별 만들기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <set>
#define INF 987654321

using namespace std;

bool isDivisible(int xNumerator, int yNumerator, int denominator) {
    return xNumerator % denominator == 0 && yNumerator % denominator == 0;
}

vector<string> solution(vector<vector<int>> line) {
    vector<string> answer;
    set<pair<int, int>> meets; 
    
    for(int i=0; i<line.size()-1; i++) {
        for(int j=i+1; j<line.size(); j++) {
            int a = line[i][0], b = line[i][1], e = line[i][2];
            int c = line[j][0], d = line[j][1], f = line[j][2];
            
            int denominator = a*d - b*c;

            if (denominator == 0) {
                continue;
            } else {
                int xNumerator = b*f - e*d;                
                int yNumerator = e*c - a*f;
                
                
                if (isDivisible(xNumerator, yNumerator, denominator)) {
                    int x = xNumerator / denominator;
                    int y = yNumerator / denominator;
                    
                    meets.insert({x, y});
                }
            }
        }
    }
    
    int minX = INF;
    int maxX = -INF;
    int minY = INF;
    int maxY = -INF;
    
    for(auto i : meets) {
        cout << i.first << "," << i.second << "\n";
        
        minX = min(i.first, minX);
        maxX = max(i.first, maxX);
        
        minY = min(i.second, minY);
        maxY = max(i.second, maxY);
    }
    
    cout << minX << "~" << maxX << "\n";
    cout << minY << "~" << maxY << "\n";

    return answer;
}